mypy/checkmember.py

```diff
@@ -371,8 +371,27 @@ def validate_super_call(node: FuncBase, mx: MemberContext) -> None:
 def analyze_type_callable_member_access(name: str, typ: FunctionLike, mx: MemberContext) -> Type:
     # Class attribute.
     # TODO super?
-    ret_type = typ.items[0].ret_type
+    item = typ.items[0]
+    ret_type = item.ret_type
     assert isinstance(ret_type, ProperType)
+
+    # The following is a hack. mypy often represents types as CallableType, where the signature of
+    # CallableType is determined by __new__ or __init__ of the type (this logic is in
+    # type_object_type). Then if we ever need the TypeInfo or an instance of the type, we fish
+    # around for the return type, e.g. in CallableType.type_object. Unfortunately, this is incorrect
+    # if __new__ returns an unrelated type, but we can kind of salvage things here using
+    # CallableType.bound_args
+    self_type: Type = typ
+    if len(item.bound_args) == 1 and item.bound_args[0]:
```

Comment:
I agree with other comments that using `bound_args` doesn't look like something we want to do. I think storing the real self type in a new attribute might be a better approach. All type operations should be updated to do something reasonable with the attribute. 

It seems like the above should also fix the issue that no error is currently reported here, and perhaps other related issues:
```
class C:
    def __new__(self) -> D:
        return D()

class D(C):
    x: int

C.x  # No Error, but fails at runtime
```
We'd use use the new attribute when looking up class attributes.